### PR TITLE
Manifest files

### PR DESCRIPTION
**Pull request by gambol99**
The current implementation uses flags embedded into the command line rather than argument which is more consistent with other manifests and kubeadm. Logging currently pushes all the logs into the host, which isn't really ideal with an ephemeral host model, users should be pushed to implement a proper remote logging stack to retain logs not ssh to boxes.

- changed the manifest to use arguments rather than option flags
- changed the logging to use journal rather than relying on host logging files

```shell
core@ip-10-250-34-21 ~ $ cat /etc/kubernetes/manifests/kube-apiserver.manifest  | grep -A 6 args
  - args:
    - --address=127.0.0.1
    - --admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,ResourceQuota,PodTolerationRestriction
    - --allow-privileged=true
    - --anonymous-auth=false
    - --apiserver-count=3
    - --authorization-mode=RBAC
```

### CODE REVIEW

**Comment by chrislovecnm** [comment]
Thanks for the PR

This is a big change :) There has been an issue in for a long time around getting the logs accessible via kubectl logs. I am kinda on the fence about this, because this is going to change what admins are use to a bunch.  No more log files ;)

I like using journal, but again big change in my mind. Let me ponder a bit...

File: pkg/flagbuilder/build_flags.go [hunk 1]
```diff
@@ -29,8 +29,18 @@ import (
 	"github.com/golang/glog"
 )
 
-// BuildFlags builds flag arguments based on "flag" tags on the structure
+// BuildFlags returns a space seperated list arguments
```
Comment (justinsb) [inline pkg/flagbuilder/build_flags.go]: We should probably mark this deprecated (just with a comment saying to prefer calling BuildFlagsList)

File: pkg/flagbuilder/build_flags.go [hunk 2]
```diff
@@ -176,15 +186,15 @@ func BuildFlags(options interface{}) (string, error) {
...
 			flags = append(flags, flag)
 		}
-		// Nothing more to do here
+
 		return utils.SkipReflection
 	}
 	err := utils.ReflectRecursive(reflect.ValueOf(options), walker)
 	if err != nil {
-		return "", err
+		return []string{}, err
```
Comment (justinsb) [inline pkg/flagbuilder/build_flags.go]: FYI, typically `nil` is basically equivalent to `[]string{}` for all practical use cases, so you can typically just `return nil, err`

**Comment by justinsb** [comment]
I agree with the change @gambol99 , except that (like @chrislovecnm) I do have concerns about backwards compatibility.

One option might be the same trick I suggested for the IAM changes https://github.com/kubernetes/kops/pull/3186#issuecomment-323095237 ... one day we'll enable a logging solution "out of the box" just like we do networking.  Enabling one of those logging solutions won't be the default when we first introduce it, but if the user enables it they get this more sensible logging.  And until then we have a feature flag.

I think the options are:

* We decide that it is OK to "break" logging for these components, which seems unlikely to me
* We log to both the local file and stdout, like we do with kube-proxy now.  No need for a flag.
* We have selective behaviour where we send it to one or both destinations, driven by a feature flag or field.

I'm inclined to suggest that we do the second option, but see if we can transform this PR into that because then it will make the 3rd option easier.

What do you think @gambol99 ?  And which logging solution are you using?

File: pkg/flagbuilder/build_flags.go (hunk 2 above)
Comment (gambol99) [inline pkg/flagbuilder/build_flags.go]: fixed :-)

File: pkg/flagbuilder/build_flags.go (hunk 1 above)
Comment (gambol99) [inline pkg/flagbuilder/build_flags.go]: added

**Comment by gambol99** [comment]
hi @justinsb / @chrislovecnm ... 

> One option might be the same trick I suggested for the IAM changes #3186 (comment) ... one day we'll enable a logging solution "out of the box" just like we do networking. Enabling one of those logging solutions won't be the default when we first introduce it, but if the user enables it they get this more sensible logging. And until then we have a feature flag.

Off topic to the PR but i'm inclined to forgo Kops managing any of those deployments; I'm not sure but I kinda feel it should only manage addon's which are functional to the cluster (a CNI or kube-dns etc) i.e. operations. Anything which is "feature based" should out of scope and be part of an application-pipeline. It reduces the conditionals, api spec, templating and things Kops need to think about.  

> I'm inclined to suggest that we do the second option, but see if we can transform this PR into that because then it will make the 3rd option easier.

Agreed!, I've reverted the changes and it will now log to both stdout and original logging file

>  And which logging solution are you using?

At present a standard ELK stack, although it lives externally and isolated to clusters

**Comment by chrislovecnm** [comment]
> . Anything which is "feature based" should out of scope and be part of an application-pipeline. It reduces the conditionals, API spec, templating and things Kops need to think about.

Yes, I agree.  The question is if logging is a base component or not.  Since there are like 40 different solutions I am thinking no.  @justinsb has been threatening to build a deployment manager. Dashboard, logging, an other stuff could go there.  

We already have some of the functionality needed in channels.  

Admittedly kops deploying other manifests has been a big discussion item.

**Comment by justinsb** [comment]
Yes, up until now the line has been whether we need a component in order to bring up the system.  Networking therefore is in.  kube2iam or IAM lockdown is not in today, but I could imagine that we might have a really locked down configuration in future where e.g. kube-controller-manager has different IAM permissions from kube-apiserver, and then maybe kube2iam would be required.  The idea of baking kube2iam in was convenient for figuring out how to incrementally lock down IAM permissions, but I think we can just make it a separate field.

Logging hasn't been one that I thought should be in, except that it would have been convenient here to incrementally adopt the change.   (And maybe for logging/diagnostics during system bootstrap).

And yes, my hope is that we can get all the addons to be a simple manifest, in which case this might become largely academic anyway.

File: nodeup/pkg/model/kube_proxy.go [hunk 3]
```diff
@@ -120,30 +120,26 @@ func (b *KubeProxyBuilder) buildPod() (*v1.Pod, error) {
...
-	glog.V(2).Infof("built kube-proxy command: %s", cmd)
-
-	command := []string{
-		"/bin/sh", "-c",
-		cmd,
-	}
+	flags = append(flags, []string{
+		"--conntrack-max-per-core=131072",
+		"--kubeconfig=/var/lib/kube-proxy/kubeconfig",
+		"--oom_score_adj=-998",
```
Comment (justinsb) [inline nodeup/pkg/model/kube_proxy.go]: Upstream added the code which did `echo -998 > /proc/$$$/oom_score_adj && ...`, so I opened https://github.com/kubernetes/kubernetes/issues/51083 to ask upstream / @vishh

**Comment by justinsb** [comment]
/lgtm

Thanks for changing the behaviour @gambol99

**Comment by gambol99** [comment]
this one needs fixing @justinsb, no need to retest yet... With the revert logging to both pipes it's causing issues which need fixing :-( ...

**Comment by justinsb** [comment]
Fair enough - there is also an upstream issue - https://github.com/kubernetes/kubernetes/issues/51128, hopefully fixed by https://github.com/kubernetes/kubernetes/pull/51144

**Comment by eparis** [comment]
/retest
merged the referenced fix

**Comment by gambol99** [comment]
sorry for the lack of update on this one .. I did hit a stumbling block trying to get them both to work at the same time ... Given the arguments are now being passed to the entrypoint the use of /bin/sh -c "<command | tee -a <logfile>" will no longer work ... I tried fiddling with ```(tee -a /var/log/kube-apiserver.log) 2>&1; /usr/local/bin/kube-apiserver``` or ```exec > >(tee -a /var/log/kube-apiserver.log) 2>&1; /usr/local/bin/kube-apiserver``` and a plethora of others!! .. The only way i got it to work was almost criminal, i write a entrypoint.sh and map onto the container by a volume and change the entrypoint to that; do a simple.  

```shell
#!/bin/sh
logfile=$1; shift 1;
$@ | tee -a ${logfile}
```
Or just having an extra option in the cluster spec that permits a selection ...

**Comment by gambol99** [comment]
@KashifSaadat ... can i get a review please

File: nodeup/pkg/model/kube_apiserver.go [hunk 4]
```diff
@@ -82,17 +81,13 @@ func (b *KubeAPIServerBuilder) Build(c *fi.ModelBuilderContext) error {
...
-			Path:        "/var/log/kube-apiserver.log",
-			Contents:    fi.NewStringResource(""),
-			Type:        nodetasks.FileType_File,
-			Mode:        s("0400"),
-			IfNotExists: true,
-		}
-		c.AddTask(t)
-	}
+	c.AddTask(&nodetasks.File{
+		Path:        "/var/log/kube-apiserver.log",
```
Comment (KashifSaadat) [inline nodeup/pkg/model/kube_apiserver.go]: Could possibly make this path a constant as you're referring to it in a few places

File: nodeup/pkg/model/kube_proxy.go [hunk 5]
```diff
@@ -120,45 +121,39 @@ func (b *KubeProxyBuilder) buildPod() (*v1.Pod, error) {
...
-		Image:           image,
-		ImagePullPolicy: v1.PullIfNotPresent,
-		Command:         command,
-		SecurityContext: &v1.SecurityContext{
-			Privileged: fi.Bool(true),
+		Name:  "kube-proxy",
+		Image: image,
+		Command: []string{
+			"/bin/sh", "-c",
+			"/usr/local/bin/kube-proxy " + strings.Join(sortedStrings(flags), " ") + " 2>&1 | /usr/bin/tee -a /var/log/kube-proxy.log",
```
Comment (KashifSaadat) [inline nodeup/pkg/model/kube_proxy.go]: Similar logic could be abstracted out to a helper function

File: pkg/apis/kops/cluster.go [hunk 6]
```diff
@@ -61,8 +61,7 @@ type ClusterSpec struct {
 	MasterPublicName string `json:"masterPublicName,omitempty"`
 	// MasterInternalName is the internal DNS name for the master nodes
 	MasterInternalName string `json:"masterInternalName,omitempty"`
-	// The CIDR used for the AWS VPC / GCE Network, or otherwise allocated to k8s
-	// This is a real CIDR, not the internal k8s network
+	// The CIDR used for the AWS VPC / GCE Network, or otherwise allocated to k8s. This is a real CIDR, not the internal k8s network
```
Comment (KashifSaadat) [inline pkg/apis/kops/cluster.go]: "NetworkCIDR is the CIDR used for.."

File: pkg/apis/kops/cluster.go [hunk 7]
```diff
@@ -144,11 +143,8 @@ type ClusterSpec struct {
 	CloudLabels map[string]string `json:"cloudLabels,omitempty"`
 	// Hooks for custom actions e.g. on first installation
 	Hooks []HookSpec `json:"hooks,omitempty"`
-	// Alternative locations for files and containers
-	// This API component is under construction, will remove this comment
-	// once this API is fully functional.
+	// Alternative locations for files and containers; the API under construction, will remove this comment once this API is fully functional.
```
Comment (KashifSaadat) [inline pkg/apis/kops/cluster.go]: Comment start with "Assets.."

File: pkg/flagbuilder/build_flags.go [hunk 8]
```diff
@@ -176,15 +187,15 @@ func BuildFlags(options interface{}) (string, error) {
...
 			flags = append(flags, flag)
 		}
-		// Nothing more to do here
+
 		return utils.SkipReflection
 	}
 	err := utils.ReflectRecursive(reflect.ValueOf(options), walker)
 	if err != nil {
-		return "", err
+		return nil, err
```
Comment (KashifSaadat) [inline pkg/flagbuilder/build_flags.go]: Could add some context to the errors being returned

**Comment by KashifSaadat** [comment]
Added a few comments, nothing big though.

Looks good to me! 👍

**Comment by chrislovecnm** [comment]
Arg ... needs a rebase again ... DOH

Or github is confused.

**Comment by gambol99** [comment]
is anything outstanding on this one? @justinsb @chrislovecnm

**Comment by chrislovecnm** [comment]
Yah just needed a rebase :)

/lgtm

**Pull request merged**